cli: avoid logging command line errors in more cases

pohly opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

"kubectl list" should print a plain text explanation ("Unknown command "list"
for kubectl. ...") without treating that multi-line error as a log message.

The previous heuristic didn't work here because the command itself was not
found. A better one is to check that command execution really started.

This is still not perfect: what if a command hasn't started logging yet or
never uses logging for its output and then returns an error, for example in
`kubectl kustomize`? To handle that case, `kubectl` now always deals
with error printing itself.

#### Which issue(s) this PR fixes:
Fixes #https://github.com/kubernetes/kubernetes/issues/107012

#### Does this PR introduce a user-facing change?
```release-note
Fix 1.23 regression: Some command line errors (for example, "kubectl list" -> "unknown command") were printed as log message with escaped line breaks instead of a multi-line plain text, which made the error harder to read.
```

pohly committed:
cli: avoid logging command line errors in more cases

"kubectl list" should print a plain text explanation ("Unknown command "list"
for kubectl. ...") without treating that multi-line error as a log message.

The previous heuristic didn't work here because the command itself was not
found. A better one is to check that command execution really started.

This is still not perfect: what if a command hasn't started logging yet or
never uses logging for its output and then returns an error? That works in
kubectl because it does its own error checking at runtime and then doesn't
return, but a more robust solution might be an explicit parameter that prevents
using klog.

pohly committed:
cli: let kubectl handle error printing

cli.Run was an attempt to elliminate error handling in Kubernetes
commands. However, it had to rely on heuristics that are not necessarily right
for all commands.

kubectl is one example which has its own error printing code that should be
used in all cases after a command failure. It now gets used also for
`--warnings-as-errors`. Previously, that caused the following message to be
logged at the end:

  E0110 16:56:01.987555  202060 run.go:120] "command failed" err="1 warning received"

Now it ends with:

 error: 1 warning received

xyz-li commented on cmd/kubectl/kubectl.go:
```diff
@@ -17,17 +17,18 @@ limitations under the License.
 	command := cmd.NewDefaultKubectlCommand()
-	code := cli.Run(command)
-	os.Exit(code)
+	if err := cli.RunNoOutput(command); err != nil {
```
RunNoErrOutput

pohly replied:
Darn. Well, now you know what my original name was before a last-minute change of mind...

pohly commented:
/retest

@xyz-li: PR should be okay now, there was just one flake in pull-kubernetes-integration.

pohly commented:
/retest

pohly commented:
/retest

xyz-li commented:
/lgtm

dims commented:
/retest

pohly commented:
/retest

pohly commented:
The `deprecated_api` test-cmd case seems to be sensitive to the output format. It's failing consistently at the moment:
```
+++ [0106 03:10:59] Testing deprecated APIs
Successful
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
has:PodSecurityPolicy is deprecated
Successful
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
error: 1 warning received
has:PodSecurityPolicy is deprecated
FAIL!
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
error: 1 warning received
has not:err="1 warning received"
424 /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../test/cmd/get.sh
```

For comparison, on master the output is:
```
+++ Running case: test-cmd.run_deprecated_api_tests 
+++ working dir: /nvme/gopath/src/k8s.io/kubernetes
+++ command: run_deprecated_api_tests
+++ [0110 16:56:01] Creating namespace namespace-1641830161-21779
namespace/namespace-1641830161-21779 created
Context "test" modified.
+++ [0110 16:56:01] Testing deprecated APIs
Successful
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
has:PodSecurityPolicy is deprecated
Successful
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
E0110 16:56:01.987555  202060 run.go:120] "command failed" err="1 warning received"
has:PodSecurityPolicy is deprecated
Successful
message:Warning: policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
No resources found
E0110 16:56:01.987555  202060 run.go:120] "command failed" err="1 warning received"
has:err="1 warning received"
+++ exit code: 0
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
No resources found
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
No resources found
+++ [0110 16:56:02] TESTS PASSED
```

pohly commented:
I think ending with `error: 1 warning received` is the cleaner output and therefore I have adapted the test case to match that.

pohly commented:
/retest

eddiezane reviewed (commented):
Can you please add an integration test to catch a regression here?

pohly commented:
> Can you please add an integration test to catch a regression here?

@eddiezane: With integration test, do you mean something like https://github.com/kubernetes/kubernetes/blob/a0dfd958d5c5e8c2bd5fd30e1d2e9031e9aba1b6/test/cmd/get.sh#L411-L428?

pohly committed:
kubectl: add integration test for result reporting

This currently covers two cases:
- "kubectl list" (the regression from https://github.com/kubernetes/kubernetes/issues/107012)
- "kubectl get pods/no-such-pod" (no particular reason except that the output
should be deterministic)

In contrast to some other tests that check for strings inside the
output (run_deprecated_api_tests) or compare after
sorting (run_kubectl_version_tests), stdout, stderr and the return code must
match exactly.

This ensures that there is no extra, unexpected output and that the right
output stream is used.

pohly commented:
I've added such a test for "kubectl list" and "kubectl get pods/no-such-pods". See latest commit message for details.

There's nothing special about "kubectl get pods/no-such-pods", but it can be tested because the output should be deterministic. More test cases would be even better, to catch regressions *before* they occur... :sweat_smile:

pohly commented:
/retest

soltysh commented on cmd/kubectl/kubectl.go:
```diff
@@ -17,17 +17,18 @@ limitations under the License.
-
 	"k8s.io/component-base/cli"
 	"k8s.io/kubectl/pkg/cmd"
+	"k8s.io/kubectl/pkg/cmd/util"
```
Nit: `cmdutil` is the preferred alias for these to make it more explicit.

soltysh reviewed (approved):
/lgtm
/approve

soltysh commented:
/priority important-longterm
/kind regression
/triage accepted
